## Data Ingestion and Data Preprocessing (Task 1)


In [1]:
# import needed libraries for data wrangling
import pandas as pd
import re

In [4]:
# Load dataset as  DataFrame
df = pd.read_csv(r'E:\KAIM 2\KAIM 2 Week 5\Week5\data\telegram_data.csv')

In [5]:
df

Channel Title Channel Username     ID  \
0     SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14841   
1     SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14840   
2     SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14839   
3     SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14838   
4     SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14837   
...                 ...              ...    ...   
4618  SINA KIDS/ሲና ኪድስⓇ        @sinayelj    599   
4619  SINA KIDS/ሲና ኪድስⓇ        @sinayelj    598   
4620  SINA KIDS/ሲና ኪድስⓇ        @sinayelj    197   
4621  SINA KIDS/ሲና ኪድስⓇ        @sinayelj    182   
4622  SINA KIDS/ሲና ኪድስⓇ        @sinayelj      1   

                                                Message  \
0                                                   NaN   
1                                                   NaN   
2     ለኮንዶሚኒየም ለጠባብ ቤቶች ገላግሌ የሆነ ከንፁህ የሲልከን ጥሬ እቃ የተ...   
3                                                   NaN   
4                                                   NaN   
...                                                 ...   
4618                                                NaN   
4619                 Baby potty\n0905707448\n0945097042   
4620                                                NaN   
4621  ውድ የሲና ኪድስ ደምበኞች በድጋሚ ገብቷል \nየመዋኛ ገንዳ ትልቅ ሳይዝ ...   
4622                                                NaN   

                           Date                  Media Path  
0     2024-09-26 12:24:51+00:00  photos\@sinayelj_14841.jpg  
1     2024-09-26 12:24:51+00:00  photos\@sinayelj_14840.jpg  
2     2024-09-26 12:24:51+00:00  photos\@sinayelj_14839.jpg  
3     2024-09-26 12:24:12+00:00  photos\@sinayelj_14838.jpg  
4     2024-09-26 12:24:12+00:00  photos\@sinayelj_14837.jpg  
...                         ...                         ...  
4618  2021-04-16 18:12:10+00:00    photos\@sinayelj_599.jpg  
4619  2021-04-16 18:12:10+00:00    photos\@sinayelj_598.jpg  
4620  2020-11-26 18:30:52+00:00                         NaN  
4621  2020-11-10 06:59:31+00:00    photos\@sinayelj_182.jpg  
4622  2020-07-24 10:50:43+00:00                         NaN  

[4623 rows x 6 columns]

In [7]:
df.columns

Index(['Channel Title', 'Channel Username', 'ID', 'Message', 'Date',
       'Media Path'],
      dtype='object')

In [8]:
df.info

<bound method DataFrame.info of           Channel Title Channel Username     ID  \
0     SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14841   
1     SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14840   
2     SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14839   
3     SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14838   
4     SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14837   
...                 ...              ...    ...   
4618  SINA KIDS/ሲና ኪድስⓇ        @sinayelj    599   
4619  SINA KIDS/ሲና ኪድስⓇ        @sinayelj    598   
4620  SINA KIDS/ሲና ኪድስⓇ        @sinayelj    197   
4621  SINA KIDS/ሲና ኪድስⓇ        @sinayelj    182   
4622  SINA KIDS/ሲና ኪድስⓇ        @sinayelj      1   

                                                Message  \
0                                                   NaN   
1                                                   NaN   
2     ለኮንዶሚኒየም ለጠባብ ቤቶች ገላግሌ የሆነ ከንፁህ የሲልከን ጥሬ እቃ የተ...   
3                                                   NaN   
4                                                   NaN   
.

In [9]:
# check for any missed values
print("Checking for NaN values in the 'Message' column:")
nan_count = df['Message'].isnull().sum()
print(f"Number of NaN values in 'Message' column: {nan_count}")

Checking for NaN values in the 'Message' column:
Number of NaN values in 'Message' column: 3268


The above code checks for missing values (NaN) in the 'Message' column of a DataFrame named df. It first prints a message indicating that it's checking for NaN values, and then calculates the total count of NaN values in the 'Message' column using the isnull() method followed by sum(). The result, which shows that there are 3,268 missing values, is printed to the console. This indicates that a significant portion of the data in the 'Message' column is incomplete, which may require data cleaning or handling before any further analysis or processing can be performed.

In [10]:
df = df.dropna(subset=['Message'])

# Print the shape of the dataset after dropping NaN values in the "Message" column
print(f"Dataset shape after dropping NaN values in 'Message' column: {df.shape}")

Dataset shape after dropping NaN values in 'Message' column: (1355, 6)


The above code modifies the DataFrame df by removing any rows that contain NaN values in the 'Message' column using the dropna() method. After executing this operation, it prints the shape of the updated DataFrame, which is now (1355, 6). This output indicates that the DataFrame has 1,355 rows and 6 columns remaining after dropping the rows with missing values in the 'Message' column. This reduction in the number of rows demonstrates the impact of the missing data on the dataset's size and suggests that further data analysis can now be conducted on a cleaner dataset.

In [11]:
df.head()

Channel Title Channel Username     ID  \
2   SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14839   
6   SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14835   
9   SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14832   
16  SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14825   
26  SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14815   

                                              Message  \
2   ለኮንዶሚኒየም ለጠባብ ቤቶች ገላግሌ የሆነ ከንፁህ የሲልከን ጥሬ እቃ የተ...   
6   Baby romper   ከላዩ ፈር ውስጡ ኮተን የሆነ\n\n2000 birr\...   
9   Chicco 5 in 1 mama bag\nኦሪጅናል ማቴሪያል\nበሳይዙ ትልቅ\...   
16  aiebao HIP SEAT CARRIER\nምቹ ጠንካራ የልጆች ማዘያ\nበተለ...   
26  Kids mat  \n\n💥4100 ብር\n\nSize 1.80*2m\n\nfree...   

                         Date                  Media Path  
2   2024-09-26 12:24:51+00:00  photos\@sinayelj_14839.jpg  
6   2024-09-26 12:24:11+00:00  photos\@sinayelj_14835.jpg  
9   2024-09-26 12:20:22+00:00  photos\@sinayelj_14832.jpg  
16  2024-09-26 12:19:20+00:00  photos\@sinayelj_14825.jpg  
26  2024-09-25 06:15:53+00:00  photos\@sinayelj_14815.jpg

In [12]:
message_df=df['Message']
message_df

2       ለኮንዶሚኒየም ለጠባብ ቤቶች ገላግሌ የሆነ ከንፁህ የሲልከን ጥሬ እቃ የተ...
6       Baby romper   ከላዩ ፈር ውስጡ ኮተን የሆነ\n\n2000 birr\...
9       Chicco 5 in 1 mama bag\nኦሪጅናል ማቴሪያል\nበሳይዙ ትልቅ\...
16      aiebao HIP SEAT CARRIER\nምቹ ጠንካራ የልጆች ማዘያ\nበተለ...
26      Kids mat  \n\n💥4100 ብር\n\nSize 1.80*2m\n\nfree...
                              ...                        
4611    ምቹ የልጆች ፖፖ የሶፍት ማስቀመጫ ያለው 1000ብር\nFree deliver...
4613    ለልጆች ምቹ እና ኮንፈርተብል ማት \nከላይ ማጫወቻ ያለዉ ከተፈለገ ማጫወ...
4615    Baby potty \nበባትሪ የሚሰራ ድምፅ ያለው\nየልጆች ፖፖ 899 ብር...
4619                   Baby potty\n0905707448\n0945097042
4621    ውድ የሲና ኪድስ ደምበኞች በድጋሚ ገብቷል \nየመዋኛ ገንዳ ትልቅ ሳይዝ ...
Name: Message, Length: 1355, dtype: object

In [13]:

# Example of your DataFrame
# df = pd.DataFrame({'Message': ['💥3pcs silicon brush spatulas...', '💥Mandoline Slicer...', ...]})

# Define a function to remove emojis
def remove_emojis(text):
    emoji_pattern = re.compile(
        "[" 
        "\U0001F600-\U0001F64F"  # emoticons
        "\U0001F300-\U0001F5FF"  # symbols & pictographs
        "\U0001F680-\U0001F6FF"  # transport & map symbols
        "\U0001F700-\U0001F77F"  # alchemical symbols
        "\U0001F780-\U0001F7FF"  # Geometric Shapes Extended
        "\U0001F800-\U0001F8FF"  # Supplemental Arrows-C
        "\U0001F900-\U0001F9FF"  # Supplemental Symbols and Pictographs
        "\U0001FA00-\U0001FA6F"  # Chess Symbols
        "\U0001FA70-\U0001FAFF"  # Symbols and Pictographs Extended-A
        "\U00002702-\U000027B0"  # Dingbats
        "\U000024C2-\U0001F251" 
        "]+", 
        flags=re.UNICODE
    )
    return emoji_pattern.sub(r'', text)

# Apply the function to the 'Message' column
df['Message'] = df['Message'].apply(remove_emojis)

# Display the updated DataFrame
print(df.head())


        Channel Title Channel Username     ID  \
2   SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14839   
6   SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14835   
9   SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14832   
16  SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14825   
26  SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14815   

                                              Message  \
2   ለኮንዶሚኒየም ለጠባብ ቤቶች ገላግሌ የሆነ ከንፁህ የሲልከን ጥሬ እቃ የተ...   
6   Baby romper   ከላዩ ፈር ውስጡ ኮተን የሆነ\n\n2000 birr\...   
9   Chicco 5 in 1 mama bag\nኦሪጅናል ማቴሪያል\nበሳይዙ ትልቅ\...   
16  aiebao HIP SEAT CARRIER\nምቹ ጠንካራ የልጆች ማዘያ\nበተለ...   
26  Kids mat  \n\n4100 ብር\n\nSize 1.80*2m\n\nfree ...   

                         Date                  Media Path  
2   2024-09-26 12:24:51+00:00  photos\@sinayelj_14839.jpg  
6   2024-09-26 12:24:11+00:00  photos\@sinayelj_14835.jpg  
9   2024-09-26 12:20:22+00:00  photos\@sinayelj_14832.jpg  
16  2024-09-26 12:19:20+00:00  photos\@sinayelj_14825.jpg  
26  2024-09-25 06:15:53+00:00  photos\@sinayelj_14815.jpg  


C:\Users\you\AppData\Local\Temp\ipykernel_260\1963313681.py:25: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Message'] = df['Message'].apply(remove_emojis)


In [14]:
df.to_csv(r'E:\KAIM 2\KAIM 2 Week 5\Week5\data\cleaned_data.csv')

In [15]:
df.describe

<bound method NDFrame.describe of           Channel Title Channel Username     ID  \
2     SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14839   
6     SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14835   
9     SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14832   
16    SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14825   
26    SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14815   
...                 ...              ...    ...   
4611  SINA KIDS/ሲና ኪድስⓇ        @sinayelj    674   
4613  SINA KIDS/ሲና ኪድስⓇ        @sinayelj    640   
4615  SINA KIDS/ሲና ኪድስⓇ        @sinayelj    605   
4619  SINA KIDS/ሲና ኪድስⓇ        @sinayelj    598   
4621  SINA KIDS/ሲና ኪድስⓇ        @sinayelj    182   

                                                Message  \
2     ለኮንዶሚኒየም ለጠባብ ቤቶች ገላግሌ የሆነ ከንፁህ የሲልከን ጥሬ እቃ የተ...   
6     Baby romper   ከላዩ ፈር ውስጡ ኮተን የሆነ\n\n2000 birr\...   
9     Chicco 5 in 1 mama bag\nኦሪጅናል ማቴሪያል\nበሳይዙ ትልቅ\...   
16    aiebao HIP SEAT CARRIER\nምቹ ጠንካራ የልጆች ማዘያ\nበተለ...   
26    Kids mat  \n\n4100 ብር\n\nSize 1.80*2m\n\nfree ...   

In [16]:
df

Channel Title Channel Username     ID  \
2     SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14839   
6     SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14835   
9     SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14832   
16    SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14825   
26    SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14815   
...                 ...              ...    ...   
4611  SINA KIDS/ሲና ኪድስⓇ        @sinayelj    674   
4613  SINA KIDS/ሲና ኪድስⓇ        @sinayelj    640   
4615  SINA KIDS/ሲና ኪድስⓇ        @sinayelj    605   
4619  SINA KIDS/ሲና ኪድስⓇ        @sinayelj    598   
4621  SINA KIDS/ሲና ኪድስⓇ        @sinayelj    182   

                                                Message  \
2     ለኮንዶሚኒየም ለጠባብ ቤቶች ገላግሌ የሆነ ከንፁህ የሲልከን ጥሬ እቃ የተ...   
6     Baby romper   ከላዩ ፈር ውስጡ ኮተን የሆነ\n\n2000 birr\...   
9     Chicco 5 in 1 mama bag\nኦሪጅናል ማቴሪያል\nበሳይዙ ትልቅ\...   
16    aiebao HIP SEAT CARRIER\nምቹ ጠንካራ የልጆች ማዘያ\nበተለ...   
26    Kids mat  \n\n4100 ብር\n\nSize 1.80*2m\n\nfree ...   
...                                                 ...   
4611  ምቹ የልጆች ፖፖ የሶፍት ማስቀመጫ ያለው 1000ብር\nFree deliver...   
4613  ለልጆች ምቹ እና ኮንፈርተብል ማት \nከላይ ማጫወቻ ያለዉ ከተፈለገ ማጫወ...   
4615  Baby potty \nበባትሪ የሚሰራ ድምፅ ያለው\nየልጆች ፖፖ 899 ብር...   
4619                 Baby potty\n0905707448\n0945097042   
4621  ውድ የሲና ኪድስ ደምበኞች በድጋሚ ገብቷል \nየመዋኛ ገንዳ ትልቅ ሳይዝ ...   

                           Date                  Media Path  
2     2024-09-26 12:24:51+00:00  photos\@sinayelj_14839.jpg  
6     2024-09-26 12:24:11+00:00  photos\@sinayelj_14835.jpg  
9     2024-09-26 12:20:22+00:00  photos\@sinayelj_14832.jpg  
16    2024-09-26 12:19:20+00:00  photos\@sinayelj_14825.jpg  
26    2024-09-25 06:15:53+00:00  photos\@sinayelj_14815.jpg  
...                         ...                         ...  
4611  2021-04-28 19:12:19+00:00    photos\@sinayelj_674.jpg  
4613  2021-04-20 11:16:59+00:00    photos\@sinayelj_640.jpg  
4615  2021-04-16 18:27:22+00:00    photos\@sinayelj_605.jpg  
4619  2021-04-16 18:12:10+00:00    photos\@sinayelj_598.jpg  
4621  2020-11-10 06:59:31+00:00    photos\@sinayelj_182.jpg  

[1355 rows x 6 columns]

In [18]:
# label the message
import re

def label_message_utf8_with_birr(message):
    # Split the message at the first occurrence of '\n'
    if '\n' in message:
        first_line, remaining_message = message.split('\n', 1)
    else:
        first_line, remaining_message = message, ""
    
    labeled_tokens = []
    
    # Tokenize the first line
    first_line_tokens = re.findall(r'\S+', first_line)
    
    # Label the first token as B-PRODUCT and the rest as I-PRODUCT
    if first_line_tokens:
        labeled_tokens.append(f"{first_line_tokens[0]} B-PRODUCT")  # First token as B-PRODUCT
        for token in first_line_tokens[1:]:
            labeled_tokens.append(f"{token} I-PRODUCT")  # Remaining tokens as I-PRODUCT
    
    # Process the remaining message normally
    if remaining_message:
        lines = remaining_message.split('\n')
        for line in lines:
            tokens = re.findall(r'\S+', line)  # Tokenize each line while considering non-ASCII characters
            
            for token in tokens:
                # Check if token is a price (e.g., 500 ETB, $100, or ብር)
                if re.match(r'^\d{10,}$', token):
                    labeled_tokens.append(f"{token} O")  # Label as O for "other" or outside of any entity
                elif re.match(r'^\d+(\.\d{1,2})?$', token) or 'ETB' in token or 'ዋጋ' in token or '$' in token or 'ብር' in token:
                    labeled_tokens.append(f"{token} I-PRICE")
                # Check if token could be a location (e.g., cities or general location names)
                elif any(loc in token for loc in ['Addis Ababa', 'ለቡ', 'ለቡ መዳህኒዓለም', 'መገናኛ', 'ቦሌ', 'ሜክሲኮ']):
                    labeled_tokens.append(f"{token} I-LOC")
                # Assume other tokens are part of a product name or general text
                else:
                    labeled_tokens.append(f"{token} O")
    
    return "\n".join(labeled_tokens)

# Apply the updated function to the non-null messages
df['Labeled_Message'] = df['Message'].apply(label_message_utf8_with_birr)

# Display the updated DataFrame
df.head()


C:\Users\you\AppData\Local\Temp\ipykernel_260\3418764910.py:44: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Labeled_Message'] = df['Message'].apply(label_message_utf8_with_birr)


Channel Title Channel Username     ID  \
2   SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14839   
6   SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14835   
9   SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14832   
16  SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14825   
26  SINA KIDS/ሲና ኪድስⓇ        @sinayelj  14815   

                                              Message  \
2   ለኮንዶሚኒየም ለጠባብ ቤቶች ገላግሌ የሆነ ከንፁህ የሲልከን ጥሬ እቃ የተ...   
6   Baby romper   ከላዩ ፈር ውስጡ ኮተን የሆነ\n\n2000 birr\...   
9   Chicco 5 in 1 mama bag\nኦሪጅናል ማቴሪያል\nበሳይዙ ትልቅ\...   
16  aiebao HIP SEAT CARRIER\nምቹ ጠንካራ የልጆች ማዘያ\nበተለ...   
26  Kids mat  \n\n4100 ብር\n\nSize 1.80*2m\n\nfree ...   

                         Date                  Media Path  \
2   2024-09-26 12:24:51+00:00  photos\@sinayelj_14839.jpg   
6   2024-09-26 12:24:11+00:00  photos\@sinayelj_14835.jpg   
9   2024-09-26 12:20:22+00:00  photos\@sinayelj_14832.jpg   
16  2024-09-26 12:19:20+00:00  photos\@sinayelj_14825.jpg   
26  2024-09-25 06:15:53+00:00  photos\@sinayelj_14815.jpg   

                                      Labeled_Message  
2   ለኮንዶሚኒየም B-PRODUCT\nለጠባብ I-PRODUCT\nቤቶች I-PROD...  
6   Baby B-PRODUCT\nromper I-PRODUCT\nከላዩ I-PRODUC...  
9   Chicco B-PRODUCT\n5 I-PRODUCT\nin I-PRODUCT\n1...  
16  aiebao B-PRODUCT\nHIP I-PRODUCT\nSEAT I-PRODUC...  
26  Kids B-PRODUCT\nmat I-PRODUCT\n4100 I-PRICE\nብ...

The above code defines a function, label_message_utf8_with_birr, designed to label tokens in messages based on certain criteria, particularly focusing on product names, prices, and locations. The function first splits each message at the first newline character to separate the initial product name from any additional information. 

It uses regular expressions to tokenize the first line, labeling the first token as "B-PRODUCT" and subsequent tokens as "I-PRODUCT." For the remaining message, it further tokenizes each line, labeling tokens as "I-PRICE" if they represent prices (including specific currency indicators), "I-LOC" for recognized location names, and "O" for other tokens.

 After defining this function, the code applies it to the 'Message' column of the DataFrame df, creating a new column 'Labeled_Message' that contains the labeled output. 
 
 Finally, the updated DataFrame is displayed with df.head(), showcasing the first few rows of the DataFrame, including the newly labeled messages, which provides insight into the structured representation of the data for potential further analysis or machine learning tasks.

In [24]:
# Save the updated labeled dataset to a file in CoNLL format
labeled_data_birr_path = r'E:\KAIM 2\KAIM 2 Week 5\Week5\data\labeled_telegram_product_price_location.txt'
with open(labeled_data_birr_path, 'w', encoding='utf-8') as f:
    for index, row in df.iterrows():
        f.write(f"{row['Labeled_Message']}\n\n")

In [28]:
# Attempting to read the file with utf-8 encoding
try:
    with open(r'E:\KAIM 2\KAIM 2 Week 5\Week5\data\labeled_telegram_product_price_location.txt', 'r', encoding='utf-8') as file:
        lines = file.readlines()
    for line in lines:
        print(line.strip())

except UnicodeDecodeError as e:
    print(f"UnicodeDecodeError: {e}")
    # You can try another encoding if utf-8 fails
    with open(r'E:\KAIM 2\KAIM 2 Week 5\Week5\data\labeled_telegram_product_price_location.txt', 'r', encoding='latin-1') as file:
        lines = file.readlines()
    for line in lines:
        print(line.strip())

ለኮንዶሚኒየም B-PRODUCT
ለጠባብ I-PRODUCT
ቤቶች I-PRODUCT
ገላግሌ I-PRODUCT
የሆነ I-PRODUCT
ከንፁህ I-PRODUCT
የሲልከን I-PRODUCT
ጥሬ I-PRODUCT
እቃ I-PRODUCT
የተሰራ I-PRODUCT
የልጆች I-PRODUCT
ማጠቢያ I-PRODUCT
ምስሉ I-PRODUCT
ላይ I-PRODUCT
እንደሚያዩት I-PRODUCT
መታጠፍ I-PRODUCT
መዘርጋት I-PRODUCT
የሚችል I-PRODUCT
3350ብር I-PRODUCT
ይደውሉልን I-PRODUCT
እርሶ I-PRODUCT
መምጣት I-PRODUCT
ባይመቾ I-PRODUCT
እኛ I-PRODUCT
ያሉበት I-PRODUCT
ድረስ I-PRODUCT
እናደርስሎታለን I-PRODUCT
ስልክ I-PRODUCT
0905707448 I-PRODUCT
0909003864 I-PRODUCT
ሲና I-PRODUCT
የተመረጡና I-PRODUCT
ጥራታቸውን I-PRODUCT
የጠበቁ I-PRODUCT
የልጆች I-PRODUCT
እቃ I-PRODUCT
አስመጪ I-PRODUCT
0909003864 O
0905707448 O
እቃ O
ለማዘዝ O
ከስር O
ያለውን O
ሊንኮች O
በመጫን O
ማዘዝ O
ትችላላቹ O
@sinasinaye O
@sinayalj2 O
አድራሻ O
1⃣ቁጥር1 O
ገርጂ O
ኢምፔሪያል O
ከሳሚ O
ህንፃ O
ጎን O
አልፎዝ O
ፕላዛ O
ግራውንድ O
ላይ O
እንደገቡ O
ያገኙናል O
2⃣ቁጥር2 O
4ኪሎ O
ቅድስት O
ስላሴ O
ህንፃ O
ማለትም O
ከብልፅግና O
ዋናፅፈት O
ቤት O
ህንፃ O
በስተ O
ቀኝ O
ባለው O
አስፓልት O
20ሜትር O
ዝቅ O
እንዳሉ O
ሀበሻ O
ኮፊ O
የሚገኝበት O
ቀይ O
ሸክላ O
ህንፃ O
2ተኛ O
ፎቅ O
ላይ O
ያገኙናል O
3⃣ቁጥር3 O
ብስራተ O
ገብርኤል I-PRICE
ላፍቶ O
ሞል O
መግቢያው O
ፊት O
ለፊት 

The above code attempts to read a text file located at a specified path, handling potential encoding issues that may arise during the process. 

Initially, it tries to open the file using UTF-8 encoding, which is commonly used for text files. If a UnicodeDecodeError occurs—indicating that the file contains characters not supported by UTF-8—the code catches this exception and attempts to read the file again using latin-1 encoding as a fallback. 

This approach ensures that the file is read correctly regardless of its encoding, and any lines read from the file are printed to the console after stripping leading or trailing whitespace. The inclusion of error handling allows for robust file reading, accommodating various text encodings.